lib/propolis/src/vmm/time_adjust.rs

```diff
@@ -0,0 +1,550 @@
+///
+/// boot_hrtime_delta = uptime + migrate_delta
+pub fn calc_boot_hrtime_delta(
+    uptime: Duration,
+    migrate_delta: Duration,
+) -> Result<Duration, TimeAdjustError> {
+    match uptime.checked_add(migrate_delta) {
+        Some(v) => Ok(v),
+        None => {
+            Err(TimeAdjustError::TimeDeltaOverflow { uptime, migrate_delta })
+        }
+    }
+}
+
+/// Calculate the new boot_hrtime for the guest.
+///
+/// The boot_hrtime is the hrtime of when a VM booted on the current host. In
+/// the case of live migration, this VM did not boot on this host, so we need
+/// to adjust to the boot_hrtime to be the "effective boot_hrtime": that is,
+/// what the hrtime of this host would have been when this VM booted.
+///
+/// To do so, we need several pieces of information:
+/// - the current hrtime of this host
+/// - the uptime of the VM
+/// - the migration time delta
+///
+/// And we can fix up the boot_hrtime as follows:
+///     boot_hrtime = cur_hrtime - (vm_uptime_ns + wallclock_delta)
+///
+/// The `vm_uptime_ns + wallclock_delta` term is `total_delta` here.
+///
+/// Note: It is possible for the boot_hrtime to be negative, in the case that
+/// the target host has a smaller uptime than the guest. This is okay: hrtime_t
+/// is signed, and the boot_hrtime is used by bhyve as a normalization value for
+/// device timers.
+///
+pub fn calc_boot_hrtime(
+    total_delta: Duration,
+    dst_hrtime: Duration,
+) -> Result<Hrtime, TimeAdjustError> {
+    let dst_hrt_ns: i128 = dst_hrtime.as_nanos() as i128;
+    let adjusted_bhrt: i128;
+    match dst_hrt_ns.checked_sub_unsigned(total_delta.as_nanos()) {
+        Some(v) => {
+            adjusted_bhrt = v;
+        }
+        None => {
+            return Err(TimeAdjustError::BootHrtimeOverflow {
+                desc: "dst_hrtime - total_delta".to_string(),
+                total_delta,
+                dst_hrtime,
+            });
+        }
+    }
+
+    if (adjusted_bhrt < i64::MIN as i128) || (adjusted_bhrt > i64::MAX as i128)
+    {
+        return Err(TimeAdjustError::BootHrtimeOverflow {
+            desc: "boot_hrtime to i64".to_string(),
+            total_delta,
+            dst_hrtime,
+        });
+    }
+
+    Ok(adjusted_bhrt as Hrtime)
+}
+
+/// Calculate the adjustment needed for the guest TSC.
+///
+/// ticks = (migrate_delta ns * guest_hz hz) / `NS_PER_SEC`
+pub fn calc_tsc_delta(
+    migrate_delta: Duration,
+    guest_hz: u64,
+) -> Result<u64, TimeAdjustError> {
+    assert_ne!(guest_hz, 0);
+
+    let delta_ns: u128 = migrate_delta.as_nanos();
+    let mut tsc_adjust: u128 = 0;
+
+    let upper: u128;
+    if let Some(v) = delta_ns.checked_mul(guest_hz as u128) {
+        upper = v;
+    } else {
+        return Err(TimeAdjustError::TscAdjustOverflow {
+            desc: "migrate_delta * guest_hz".to_string(),
+            migrate_delta,
+            guest_hz,
+            tsc_adjust,
+        });
+    }
+
+    tsc_adjust = upper / NS_PER_SEC;
```

Comment:
Use `checked_div`, since we're being so careful with math everywhere else?

Reply:
done in cdbe3ef